### PR TITLE
add tt dev environment, add tt compat layer

### PR DESCRIPTION
**Pull request by ftelnov**
It is needed to test compatibility with tarantool environment, and ensure it correctly sets the `__serialize` marker.

It also simplifies the dev process for tarantool users.

### CODE REVIEW

File: src/tt_compat.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * Internally it loads json library, then extracts the "null" value from it and stores it in a registry.
+*/
+static void tt_lua_createnull(lua_State *L) {
+    luaL_register(L, "json", 0);
+    lua_pushlightuserdata(L, (void *)&TT_NULL_ADDR);
+    lua_getfield(L, -2, "null");
+    lua_settable(L, LUA_REGISTRYINDEX);
+}
+
+void tt_lua_pushnull(lua_State *L) {
```
Comment (godzie44) [inline src/tt_compat.cpp:22]: how about add `inline` here?

Comment (godzie44) [inline src/tt_compat.cpp]: how about add `inline` here and move to .h?

File: src/tt_compat.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  lua_pushlightuserdata(L, (void *)&TT_NULL_ADDR);
+  lua_gettable(L, LUA_REGISTRYINDEX);
+}
+
+void tt_compat_init(lua_State *L) {
+  tt_lua_createnull(L);
+}
+
+void tt_lua_push_serialize_mt(lua_State *L, char* marker) {
```
Comment (godzie44) [inline src/tt_compat.cpp]: mb `inline` this fn as well

File: src/tt_compat.cpp (hunk 1 above)
Comment (ftelnov) [reply src/tt_compat.cpp]: I would rather not doing it. It relies on the internal variable(some random address that is used to add extracted NULL to the global LUA_REGISTRY), so inlining that would mean exposing that internal variable as well.

Anyway, why do we even care about manual inlining? That's pretty small program, and we can even enable LTO to allow compiler making yet another round of optimization in hope that it would eliminate function call itself.

In other words, I would prefer to delegate such a manual labor to the machine. Let's just enable the LTO, and with the bench inspect the gain.

File: src/tt_compat.cpp (hunk 2 above)
Comment (ftelnov) [reply src/tt_compat.cpp]: Same as above - let's focus on that discussion.

File: src/tt_compat.cpp (hunk 1 above)
Comment (godzie44) [reply src/tt_compat.cpp]: ok, you can use a macros instead, or move this code in luasimdjson.cpp

File: src/tt_compat.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
+#include <lua.hpp>
+#include <lauxlib.h>
+#include <string>
+#include "tt_compat.h"
+
+char* SER_MARKER_MAP = (char*)"map";
+char* SER_MARKER_SEQ = (char*)"seq";
+
+// Variable is needed to avoid collisions in LUA_REGISTRY.
+static const char TT_NULL_ADDR = 't';
```
Comment (godzie44) [inline src/tt_compat.cpp]: Can you explain why random choice better then a constant big number?

Comment (ftelnov) [reply src/tt_compat.cpp]: This code is actually taken from the lua manual: https://www.lua.org/pil/27.3.1.html, the method is called "bulletproof".
"The C link editor ensures that this key is unique among all libraries.".

We can also just use a complex key like "__lua_simdjson__null_addr" to achieve merely the same result(to avoid collision possibility).

File: src/tt_compat.cpp (hunk 1 above)
Comment (ftelnov) [reply src/tt_compat.cpp]: LTO doesn't help us here surprisingly, so you are totally right, we must inline it ourselves. I've reworked the code.

Comment (ftelnov) [reply src/tt_compat.cpp]: Added inlining to the serialize mt pushing as well.

**Review by godzie44 — DISMISSED** [review]

**Review by lowitea — APPROVED** [review]

**Review by godzie44 — APPROVED** [review]

**Review by Kasen — APPROVED** [review]
